src/index.ts

```diff
@@ -1,3 +1,182 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 'use-strict';
+
+import {
+  IInputAreaViewModel
+} from 'jupyter-js-input-area';
+
+import {
+  IOutputAreaViewModel
+} from 'jupyter-js-output-area';
+
+import {
+  IObservableList
+} from 'phosphor-observablelist';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+import './index.css';
+
+
+/**
+ * An enum which describes the type of cell.
+ */
+enum CellType {
+  /**
+   * The cell contains code input.
+   */
+  Code,
+
+  /**
+   * The cell contains markdown.
+   */
+  Markdown,
+
+  /**
+   * The cell contains raw text.
+   */
+  Raw
+}
+
+
+/**
+ * An enum which describes which interface property has changed.
+ */
+export
+enum CellProperty {
+  CellType,
+  Tags,
+  Deletable,
+  Mergable,
+  Splittable,
+  Rendered,
+  Marked,
+  Format
+}
+
+
+/**
+ * The arguments object emitted with the `stateChanged` signal.
+ */
+export
+interface ICellChangedArgs<T> {
+  which: CellProperty,
+  oldValue: T;
+  newValue: T;
+}
+
+
+/**
+ * The definition of a model object for a base cell.
+ */
+interface IBaseCellViewModel {
+
+  /**
+   * The type of cell.
+   */
+  cellType: CellType;
+
+  /**
+   * Tags applied to the cell.
+   */
+  tags?: IObservableList<string>;
+
+  /**
+   * Get namespaced metadata about the cell.
+   */
+  getMetadata(namespace: string) : IObservableMap<string, JSONData>;
```

Comment:
The main place where cell metadata is edited is by the cell toolbars, which is entirely a View-level aspect of the cell. I think that all of our existing cell toolbars serve to simply edit that cell metadata. Because of this the cell toobars (which will probably be provided using extensions/extension points) should probably be passed the ViewModel of the cell.

The metadata shouldn't be a single property on the ViewModel as we decided that we want to:

1) Force people to namespace their metadata
2) Enable people watch signals on their particular namespace changing in bulk.